docs/tutorial/algorithm_interfaces.rst

```diff
@@ -1,2 +1,464 @@
+Now we will take a look at how different algorithm interfaces interact with each
+other in the on-policy/off-policy pipelines. We will try modifying some
+algorithm interfaces of the previous example :doc:`./a_minimal_working_example`,
+for an illustration purpose. Then we introduce a simple off-policy training
+example and again explore its algorithm interfaces.
+
+Interface Overview
+------------------
+
+.. image:: images/algorithm_interfaces.png
+    :width: 800
+    :align: center
+    :alt: interaction between algorithm interfaces
+
+
+There are several key interfaces (highlighted in the figure above) that play
+important roles in assembling a training pipeline of an ALF algorithm, and their
+signatures are summarized below.
+
+======================================================  ===========================================  =======================================  =============================== ===============
+Interface                                               Tensor shape                                 Input                                    Output                          Which pipeline?
+======================================================  ===========================================  =======================================  =============================== ===============
+:meth:`~.AlgorithmInterface.predict_step`               :math:`[1,\ldots]`                           ``inputs``, ``state``                    ``output``, ``state``, ``info`` On/Off-policy
+:meth:`~.AlgorithmInterface.rollout_step`               :math:`[B,\ldots]`                           ``inputs``, ``state``                    ``output``, ``state``, ``info`` On/Off-policy
+:meth:`~.AlgorithmInterface.train_step`                 :math:`[B,\ldots]` or :math:`[TB,\ldots]`    ``inputs``, ``state``, ``rollout_info``  ``output``, ``state``, ``info`` Off-policy
+:meth:`~.AlgorithmInterface.calc_loss`                  :math:`[T,B,\ldots]`                         ``info``                                 ``loss_info``                   On/Off-policy
+:meth:`~.AlgorithmInterface.after_update`               :math:`[T,B,\ldots]`                         ``root_inputs``, ``info``                ``None``                        On/Off-policy
+:meth:`~.AlgorithmInterface.after_train_iter`           :math:`[T,B,\ldots]`                         ``root_inputs``, ``rollout_info``        ``None``                        On/Off-policy
+======================================================  ===========================================  =======================================  =============================== ===============
+
+:math:`T` is :attr:`.TrainerConfig.unroll_length` for on-policy algorithms and
+:attr:`.TrainerConfig.mini_batch_length` for off-policy algorithms. :math:`B` is
+:attr:`.create_environment.num_parallel_environments` for on-policy algorithms and
+:attr:`.TrainerConfig.mini_batch_size` for off-policy algorithms. We can see that
```

Comment:
For rollout_step, B is always create_environment.num_parallel_environments

Reply:
updated